### ISSUE TITLE
Isolation Forest as possibility?

### ISSUE THREAD

**Issue by andrewm4894** [davisking/dlib#2686]
Hello - i'm wondering if there might already be most of the ingredients in dlib to implement an [isolation forest](https://en.wikipedia.org/wiki/Isolation_forest#Algorithm)?

I see [`random_forest_regression_trainer`](http://dlib.net/dlib/random_forest/random_forest_regression_abstract.h.html#random_forest_regression_trainer) so am guessing maybe would be building blocks from that that could maybe be used.

Background here is that we (Netdata) are currently using [kmeans](https://github.com/netdata/netdata/blob/master/ml/KMeans.cc#L4) for some unsupervised anomaly detection and i am debating if we try move next (to support broader feature repesentastions that can take account of things like time of day and other features more easily) to something like [one class svm](http://dlib.net/ml.html#svm_one_class_trainer) or maybe even see about contributing some variant of isolation forest potentially.

Ultimatly i want us to use autoencoders but unclear or unsure if we'd ever be able to do that efficiently enough in our monitoring agent, so was thinking isolation forest or one class svm could be next slighlty more advanced model to try.

**Comment by davisking** [comment]
Yeah the isolation forest idea looks easy to implement.  You could probably do that without much trouble.

I do think things like radial basis function kernel models that do some kind of density estimation are better than this kind of approach though :)  E.g. http://dlib.net/one_class_classifiers_ex.cpp.html, especially with the empirical kernel map they are ultra fast.

All of these things come down to some thoughtful way to make the vector space the density estimator (which is what all these things ultimately are, e.g. they go "is this in a region that has hardly any other examples in it?").  And there isn't any way to build that without thoughtful domain insight.  Since for example, there are always variable that, in your particular do main, would produce anomaly samples but you say "oh but we don't care about that".  

For example, say you were looking for people doing suspicious things outside some building.  Stuff like time of day, do they appear to be trying to be hidden, are they looking over their shoulder, etc are relevant features.  But their hair color, the temperature in a city on the other side of the world, or how close Neptune is to Jupiter would be irrelevant features.  And including them would create spurious anomalies.

Sometimes it's hard to make up the right features, but easier to say which samples are "similar".  In such cases you can sometimes use some kind of manifold regularization method (e.g. http://dlib.net/ml.html#linear_manifold_regularizer) to generate a new feature space that encodes your domain's sense of "is that a weird sample?" in the distance between points.

**Comment by andrewm4894** [comment]
Cool thanks - yep i think i need to look into ocsvm and various kernel tricks etc. At moment our feature vecotr is just say 6 floating point numbers that are the differenced, smoothed and lagged last 6 observations of a metric. 

So with current kmeans implementation its really just comparing lastest feature vector to the 2 trained cluster centroids.

We want to extend this to have some more awareness of context like time of day etc. So i might try and see if i can get our dev's to benchmark a more extended feature vector with some time of day type features to maybe try get more towards "does this pattern look strange, for this time of day" and the existing once class svm stuff. 

At moment we are [extending](https://github.com/netdata/netdata/discussions/12763) the kmeans approach to train multiple models to sort of cover the last 24 hours or longer (so train and store model every 4 hours and then at inference check agaisnt each model as needed). So that would get us "does the latest feature vector look strange compared to trained kmenas models over the last 24 hours" (current default is just train on last 4 hours of data).

I did this blog post a while back explaining our implementation if curious (all farily basic, just wanted to start with something sort of medium complexity like kmeans, so better than a zscore but not quite as powerful as maybe some other models) : https://blog.netdata.cloud/how-netdatas-machine-learning-works/ 

I see that [ocsvm in sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDOneClassSVM.html?highlight=partial_fit#sklearn.linear_model.SGDOneClassSVM.partial_fit) also supports `partial_fit` so need to explore that too as being able to train incrementally could be very very useful for us in terms of overhead on agent etc.
